### PR TITLE
Hyperlanes

### PR DESCRIPTION
**Pull request by Somberlord**
Working hyperlanes for combat and cabal Hero.

Mainly tested with PoK hyperlanes, but should work with Phil's.
Added extra test script to add a "Get neighbors" context menu on systems, which logs neighbors in console.

Removed systems context menus on hyperlanes.

### CODE REVIEW

**Comment by darrellanderson** [comment]
Awesome, this is really great.  I have a few minor comments I'll add a code review.  Thanks!

File: TI4/Helpers/TI4_SystemHelper.ttslua [hunk 1]
```diff
@@ -1758,7 +1889,7 @@ function _addContextMenuItems(systemObject)
 end
 
 function onObjectSpawn(object)
-    if _systems[object.getGUID()] then
+    if _systems[object.getGUID()] and not _systems[object.getGUID()].hyperlane then
```
Comment (darrellanderson) [inline TI4/Helpers/TI4_SystemHelper.ttslua:1892]: nice catch

File: Tests/TestHyperlanes.ttslua [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    local function getCallWrapper(functionName)
+        helperObject = helperObject or getHelperObject()
+        if not helperObject.getVar(functionName) then error('missing ' .. helperObjectName .. '.' .. functionName) end
+        return function(parameters) return helperObject.call(functionName, parameters) end
+    end
+    return setmetatable({}, { __index = function(t, k) return getCallWrapper(k) end })
+end
+
+local _systemHelper = getHelperClient('TI4_SYSTEM_HELPER')
+
```
Comment (darrellanderson) [inline Tests/TestHyperlanes.ttslua:18]: THANK YOU for writing a test case -- an early version had a full unittest framework but that got dropped.  If the helpers ever get a rewrite I'd love to see that come back.

**Review by darrellanderson — COMMENTED** [review]
Overall looks great!  Some code comments, let me know what you think.

File: TI4/Helpers/TI4_SystemHelper.ttslua [hunk 3]
```diff
@@ -1535,6 +1593,79 @@ function hexAdjacentWormholes(params)
...
+            end
+        end
+    end
+    return result
+end
+
+--- Get all adjacent hexes, including hyperlane-connected hexes
+-- @param hex : hex encoded as string.
+-- @return table : list of neighbor hex strings.
+function hexNeighborsWithHyperlanes(hex)
```
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1650]: Some scripts (like zonehelper at least) need the existing hexNeighbors if I understand it correctly, so I couldn't replace default.

File: TI4/Helpers/TI4_SystemHelper.ttslua (hunk 1 above)
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1892]: Thanks, it was weird to be able to activate hyperlanes^^

File: Tests/TestHyperlanes.ttslua (hunk 2 above)
Comment (Somberlord) [reply Tests/TestHyperlanes.ttslua:18]: No problem, I needed it while coding anyway, so I cleaned it and made it available, always useful !

File: TI4/Helpers/TI4_SystemHelper.ttslua (hunk 3 above)
[the remark this answers is not in the document]
Comment (darrellanderson) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1650]: You may be right, the faction borders tool might assume they are directly adjacent.  As long as the Cabal Hero and PDS2 work can look at the rest on a case-by-case basis.

File: TI4/Helpers/TI4_SystemHelper.ttslua [hunk 4]
```diff
@@ -328,16 +328,46 @@ local _systems = {
...
-    ['3ccc62'] = { tile = 87, hyperlane = true },
-    ['ee4bf1'] = { tile = 88, hyperlane = true },
-    ['33e49a'] = { tile = 89, hyperlane = true },
-    ['7e7ee1'] = { tile = 90, hyperlane = true },
-    ['565980'] = { tile = 91, hyperlane = true },
+    -- Hyperlane tiles
+    -- hyperlanes is an array of 6 or 12 indices, representing hyperlane connexion. Indices are zero-based (0 to 6)
+    -- For double sided hyperlane tiles, the last 6 indices are for B-side
+    -- 0 is top right side, when the tile is rotated at -180° on Y axis (Y rotation of zero)
+    ['81a64a'] = { tile = 83, hyperlane = true, flippable = true, hyperlanes = {
```
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:335]: I will check that out

File: TI4/Helpers/TI4_SystemHelper.ttslua [hunk 5]
```diff
@@ -1535,6 +1593,79 @@ function hexAdjacentWormholes(params)
...
+-- @param scrHex (hex) : source hex
+-- @param hyperlaneSystem (system) : hyperlane system
+-- @return table : list of neighbor hex strings. Will not return hyperlanes hexes.
+function getHyperlaneNeighbors(srcHex, hyperlaneSystem)
+    local hyperlaneObj = getObjectFromGUID(hyperlaneSystem.guid)
+    local hyperlaneHex = hexFromPosition(hyperlaneObj.getPosition())
+    local hyperlaneNeighbors = hexNeighbors(hyperlaneHex)
+
+    local hyperlaneRotation = hyperlaneObj.getRotation()
+    local rotationOffset = math.floor(hyperlaneRotation.y / 60 + 0.5)
```
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1606]: Yes, will add safety, be sure that rotation is between 0 and 240 !

File: TI4/Helpers/TI4_SystemHelper.ttslua [hunk 6]
```diff
@@ -1535,6 +1593,79 @@ function hexAdjacentWormholes(params)
...
+    local connectedNeighborsIdx = hyperlaneSystem.hyperlanes[idxSrc]
+    for _, connectedNeighbor in pairs(connectedNeighborsIdx) do
+        local updatedIdx = (connectedNeighbor + rotationOffset) % 6 + 1
+        local connectedHex = hyperlaneNeighbors[updatedIdx]
+
+        if connectedHex ~= hyperlaneHex then
+            local neighborSystem = systemFromPosition(hexToPosition(connectedHex))
+            -- if the connected hex is an hyperlane, keep processing the path
+            if neighborSystem and neighborSystem.hyperlanes then
+                local allConnected = getHyperlaneNeighbors(hyperlaneHex, neighborSystem)
```
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1635]: Nice catch, infinite loop would be bad...

File: TI4/Helpers/TI4_SystemHelper.ttslua (hunk 5 above)
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1606]: Thought about this, you're absolutely right, don't know why I wanted to do that

File: TI4/Helpers/TI4_SystemHelper.ttslua [hunk 7]
```diff
@@ -1535,6 +1593,79 @@ function hexAdjacentWormholes(params)
...
+    -- get the list of hexes connected throuh the hyperlane
+    local connectedNeighborsIdx = hyperlaneSystem.hyperlanes[idxSrc]
+    for _, connectedNeighbor in pairs(connectedNeighborsIdx) do
+        local updatedIdx = (connectedNeighbor + rotationOffset) % 6 + 1
+        local connectedHex = hyperlaneNeighbors[updatedIdx]
+
+        if connectedHex ~= hyperlaneHex then
+            local neighborSystem = systemFromPosition(hexToPosition(connectedHex))
+            -- if the connected hex is an hyperlane, keep processing the path
+            if neighborSystem and neighborSystem.hyperlanes then
```
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1634]: That was made on purpose, in order to be consistent with hexNeighbors which return all neighbor hexes, wether they contain a system or not.
All calling function already check is the hex contains a system.

[the remark this answers is not in the document]
Comment (darrellanderson) [reply TI4/Helpers/TI4_SystemHelper.ttslua:1635]: Ah I see, these are only hexes and not systems.  Great.

**Comment by darrellanderson** [comment]
Looks ready, are you going to look at the _faceUp / _faceDown or would you like me to merge this now?  Any other changes coming?

File: TI4/Helpers/TI4_SystemHelper.ttslua (hunk 4 above)
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua]: Fixed, code is easier to understand, thanks

File: TI4/Helpers/TI4_SystemHelper.ttslua (hunk 5 above)
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua]: Fixed

File: TI4/Helpers/TI4_SystemHelper.ttslua (hunk 6 above)
[the remark this answers is not in the document]
Comment (Somberlord) [reply TI4/Helpers/TI4_SystemHelper.ttslua]: Fixed. Needed to remove self from neighbors too

**Comment by Somberlord** [comment]
Just did faceUp / faceDown, ready to merge !

**Pull request merged**